pysb/importers/bngl.py

```diff
@@ -0,0 +1,319 @@
+import warnings
+
+
+def _ns(tag_string):
+    """
+    Shorthand to apply XML namespace
+    """
+    return tag_string.format('{http://www.sbml.org/sbml/level3}')
+
+
+class BnglImportError(Exception):
+    pass
+
+
+class BnglBuilder(Builder):
+    """
+    Class for assembling a Model from a .bngl file. See
+    :py:func:`model_from_bngl` for further details.
+    """
+    def __init__(self, filename):
+        super(BnglBuilder, self).__init__()
+        with BngConsole(model=None) as con:
+            con.load_bngl(filename)
+            con.action('writeXML', evaluate_expressions=0)
+            self.x = xml.etree.ElementTree.parse('%s.xml' %
+                           con.base_filename).getroot().find(_ns('{0}model'))
+            self._model_env = {}
+        self.parse_bng_xml()
+
+    def _eval_in_model_env(self, expression):
+        """
+        Evaluates an expression string in the model environment
+        """
+        components = dict((c.name, c) for c in self.model.all_components())
+        self._model_env.update(components)
+
+        # Quick security check on the expression
+        if not re.match(r'^[\w\s\(\)/\+\-\._\*]*$', expression):
+            raise BnglImportError('Security check on expression "%s" failed' %
+                                  expression)
+
+        return eval(expression, {}, self._model_env)
+
+    def _parse_species(self, species_xml):
+        # Species may be empty for synthesis/degradation reaction patterns
+        if species_xml is None:
+            return []
+        # Give the bonds unique numeric IDs
+        bond_ids = {}
+        next_bond_id = 1
+        for bond in species_xml.iterfind(_ns('{0}ListOfBonds/{0}Bond')):
+            for bond_attr in bond.attrib.keys():
+                if bond_attr.startswith('site'):
+                    bond_ids.setdefault(bond.attrib[bond_attr], [])\
+                                        .append(next_bond_id)
+            next_bond_id += 1
+
+        # Create a list of monomer patterns
+        mon_pats = []
+        for mon in species_xml.iterfind(_ns('{0}ListOfMolecules/{0}Molecule')):
+            mon_name = mon.get('name')
+            mon_obj = self.model.monomers[mon_name]
+            mon_states = {}
+            for comp in mon.iterfind(_ns('{0}ListOfComponents/{0}Component')):
+                state_nm = comp.get('name')
+                bonds = comp.get('numberOfBonds')
+                if bonds == "0":
+                    mon_states[state_nm] = None
+                elif bonds == "1":
+                    bond_list = bond_ids[comp.get('id')]
+                    assert int(bonds) == len(bond_list)
+                    mon_states[state_nm] = bond_list[0]
+                elif bonds == "?":
+                    mon_states[state_nm] = WILD
+                elif bonds == "+":
+                    mon_states[state_nm] = ANY
+                else:
+                    raise BnglImportError('Bond %s on monomer %s has '
```

Comment:
Pysb handles multi-bonds just fine. The only place they cause trouble is Kappa export (at which point an exception is thrown if any exist). Is there another reason I'm missing that this is a problem here?